mantidproject/mantid
# Remove use of Qt types from presenter

This PR removes use of Qt types from the ISIS Reflectometry's `MainWindowPresenter` because Qt should only be used in the view but some use of it had crept into the presenter.

- Use of `QFileDialog` has been moved into functions in the view. The presenter calls them via the `IMessageHandler` interface.
- The save/load JSON functions have been moved into functions in the view. The presenter call them via a new `IFileHandler` interface. 
- Interfaces have been added for the `Encoder` and `Decoder` so they can be mocked. 

Note that not all Qt types are removed from the public interfaces for `IFileHandler`, `IEncoder` and `IDecoder` yet because this would require non-trivial conversion between Qt and standard types, which is beyond the scope of this PR.

**To test:**

Check that the GUI still functions and load/save still works, e.g.
- Open the ISIS Reflectometry interface
- Set instrument to INTER and in the table enter run=`13460` and angle=`0.5`
- Edit some settings, e.g. on the Experiment tab increase LambdaMin to `2.5`.
- On the Runs tab, click Process and check that the row is processed successfully (turns green).
- Save the contents of the GUI from the `Batch->Save` menu.
- Close and re-open the GUI.
- Re-load the contents using the `Batch->Load` menu.
- All of the table values and settings you changed should be restored. Note that the row state (green highlighting) will not be restored.
- Click Process and check it processes successfully again.

Fixes #27242

*This does not require release notes* because **it does not change functionality**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
